services/QuillLMS/app/lib/class_code_generator.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module ClassCodeGenerator
+  class MaxRetriesExceededError < StandardError; end
+
+  MAX_RETRY_COUNT = 2
+
+  def self.generate_unique_code
+    MAX_RETRY_COUNT.times do
+      code = generate
+      return code unless Classroom.unscoped.exists?(code:)
+    end
+    generate_fallback_and_report_error
+  end
+
+  def self.generate_fallback_and_report_error
+    ErrorNotifier.report(MaxRetriesExceededError.new)
+
+    generate_fallback
+  end
+
+  def self.noun1 = @noun1 ||= File.readlines(Rails.root.join('db/noun_1.db'))
+
+  def self.noun2 = @noun2 ||= File.readlines(Rails.root.join('db/noun_2.db'))
+
+  def self.generate = "#{sample(noun2)}-#{sample(noun1)}"
+
+  def self.generate_fallback = "#{generate}-#{rand(999)}"
```

Comment:
We may not need to worry about this, but it is at least theoretically possible for this to cause a collision if `rand(999)` returns the same thing for an already-extant class code.

I think we can probably just allow that to bubble up into some other error in the unlikely case that it happens rather than making this process more complicated.

Reply:
Your observation is correct; this is an implementation of what [we resolved during the post-mortem](https://www.notion.so/quill/Post-Mortem-Jan-13-2025-class-code-bug-17ad42e6f941802a8b9acd23905d350a). The intention is to gracefully alert developers when double-retries start to spike. At that time we will ship a product-facing change in the class code generation logic. We want to punt on the product-facing change for a while (couple years?).